Conditionally Add fulfillment_status column to wc_order_stats table for new installs

Nikschavan opened the pull request:
### Changes proposed in this Pull Request:

This PR adds a new `fulfillment_status` column to the `wc_order_stats` table for newly installed WooCommerce stores. This enhancement enables better tracking and filtering of orders based on their fulfillment status in analytics and reporting.

**Key changes:**
- Added `fulfillment_status` column to `wc_order_stats` table during fresh installations
- Integrated fulfillment status tracking in the Order Stats DataStore
- Column is only added for new installations to avoid altering existing store databases

Part of WOOPLUG-5707.

### Screenshots or screen recordings:

N/A - This is a database schema change for analytics purposes.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. **Test fresh installation:**
   - Set up a fresh WordPress installation
   - Install and activate WooCommerce with this branch
   - Check the database to confirm the `fulfillment_status` column exists in `wp_wc_order_stats` table
   - Verify the option `woocommerce_order_stats_has_fulfillment_column` is set to 'yes'

2. **Test existing installation:**
   - Update an existing WooCommerce installation to this branch
   - Verify the `fulfillment_status` column is NOT added to existing installations
   - Verify the option `woocommerce_order_stats_has_fulfillment_column` remains 'no'

3. **Test order creation:**
   - On a fresh installation with the new column, create test orders
   - Verify orders are created successfully
   - Check the `wp_wc_order_stats` table to see that `fulfillment_status` values are populated correctly

<!-- End testing instructions -->

### Testing that has already taken place:

- Tested on a fresh installation to verify column creation
- Verified existing installations are not affected
- Confirmed order stats data is populated correctly with fulfillment status

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message

Add fulfillment_status column to wc_order_stats table for new installations to enable better order fulfillment tracking in analytics.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment

</details>

bor0 commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -392,6 +393,29 @@ public static function newly_installed() {
+	 *
+	 * @return void
+	 */
+	public static function newly_installed_db_schema() {
```
Hi @Nikschavan, thanks for your work here!

Something a coworker noticed, I was wondering if we can re-use https://github.com/woocommerce/woocommerce/blob/f2b522435206b098ab1dd25b46fe8e27ba75c401/plugins/woocommerce/includes/class-wc-install.php#L46 here?

Nikschavan replied:
Hi @bor0! Thank you for reaching out. IMO, adding this to the `db_updates` will not work.

This column is needed only on fresh installs and should not be added on existing stores (More context here - pfKYZu-32T-p2#comment-3108)

The `$db_updates` are exactly the opposite of what I want to do here, i.e., they run on updates, not on fresh installs. Ideally, if this schema change were required for all sites, it would not need the `$db_updates` method; just changing the default schema would have updated the schema for all sites, [as that is taken care of by dbdelta](https://github.com/woocommerce/woocommerce/blob/96d1d7f233a8ac7f1670da3ea71a4adcfdd2f9f6/plugins/woocommerce/includes/class-wc-install.php#L1675).

For this case, I was planning to maybe move this as a feature that is default-enabled on new stores (similar to how HPOS is managed)

Just so you know, this PR is still WIP (hence in draft), so this is not formally ready for review. So PR description/title might be incomplete, but I very much appreciate early feedback. Do let me know if you have any thoughts or other recommendations.

I am still testing this along with related PRs in other repos and need more testing on fresh installs.

Nikschavan committed:
Add fulfillment_status column to wc_order_stats table for new installations

Nikschavan commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php:
```diff
@@ -663,6 +677,31 @@ protected static function get_net_total( $order ) {
+		}
+
+		$table_name    = self::get_db_table_name();
+		$column_exists = $wpdb->get_var(
+			$wpdb->prepare(
+				// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared -- Table name cannot be prepared.
+				"SHOW COLUMNS FROM `{$table_name}` LIKE %s",
+				'fulfillment_status'
+			)
+		);
```
Another option here would be to store an option in `get_order_stats_table_schema()`, if it is a new install, any check that option to decide if this column exists. But I thought in edge cases, if the option gets deleted somehow or whatever, accessing a column when it does not exist will cause queries failing - So I am sticking with a direct database check, but with caching in a static variable.

Nikschavan replied:
I have implemented the option to set option for caching

Nikschavan committed:
Add tests to verify changes to table schema based on if it's a fresh install

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
cs fixes

Nikschavan committed:
dont cache status of column

Nikschavan committed:
Switch DataStore fulfillment column check to use option instead of static variable

The static variable cache was breaking tests. Now using option-based
approach for checking if fulfillment_status column exists.

Nikschavan committed:
Revert "Switch DataStore fulfillment column check to use option instead of static variable"

This reverts commit 4d14b1066606810dba89d5dcdcd417e5d857c5bd.

Nikschavan committed:
Reapply "Switch DataStore fulfillment column check to use option instead of static variable"

This reverts commit 857226dffe7b6e404eef643226a00215049b6461.

Nikschavan committed:
Regenerate the table status option if it does not exist

Nikschavan committed:
Delete option to force regenerate status before tests

Nikschavan committed:
Delete option instead

Nikschavan commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -1232,6 +1233,48 @@ private static function should_enable_hpos_for_new_shop() {
+		$is_new_install = self::is_new_install();
+		if ( true === $is_new_install ) {
+			update_option( DataStore::OPTION_ORDER_STATS_TABLE_HAS_COLUMN_ORDER_FULFILLMENT_STATUS, 'yes', false );
+		}
```
Good call!

Nikschavan committed:
Don't set option for table column exists

The option is set by has_fulfillment_status_column if the option does not exist

Nikschavan committed:
Fix incorrect phpdoc param types

Nikschavan committed:
cs fixes

Nikschavan commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php:
```diff
@@ -663,6 +677,27 @@ protected static function get_net_total( $order ) {
 		return (float) $net_total;
 	}
 
+	/**
+	 * If the wc_order_stats table has the fulfillment_status column.
+	 *
+	 * @return boolean
+	 */
+	public static function has_fulfillment_status_column() {
+		global $wpdb;
+
+		$table_name    = self::get_db_table_name();
+		$column_exists = $wpdb->get_var(
+			$wpdb->prepare(
+				// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared -- Table name cannot be prepared.
+				"SHOW COLUMNS FROM `{$table_name}` LIKE %s",
+				'fulfillment_status'
+			)
+		);
+
+		self::$has_fulfillment_column = ! empty( $column_exists );
+		return self::$has_fulfillment_column;
+	}
```
This is resolved in following commits

Nikschavan committed:
cs fixes

Nikschavan committed:
CS fixes

Nikschavan commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -1232,6 +1233,45 @@ private static function should_enable_hpos_for_new_shop() {
+	/**
+	 * Get the wc_order_stats table schema.
+	 *
+	 * @since x.x.x
```
Updated in 450528904c4b32e394195dab2d57c405a1c66537

Nikschavan commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php:
```diff
@@ -655,18 +669,49 @@ protected static function get_num_items_sold( $order ) {
 	}
 
+	/**
+	 * If the wc_order_stats table has the fulfillment_status column.
```
Updated in 450528904c4b32e394195dab2d57c405a1c66537

Nikschavan commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php:
```diff
@@ -9,19 +9,28 @@
 	use StatsDataStoreTrait;
 
+	/**
+	 * Option name to store status of the wc_order_stats table has a column `fulfillment_status`
```
Updated in 450528904c4b32e394195dab2d57c405a1c66537

Nikschavan committed:
Set option also if the column does not exist to avoid repeat calls to database

Nikschavan commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php:
```diff
@@ -655,18 +669,49 @@ protected static function get_num_items_sold( $order ) {
+	/**
+	 * Check if the wc_order_stats table has the fulfillment_status column.
+	 *
+	 * @return boolean
+	 */
+	public static function has_fulfillment_status_column() {
+		$column_status = get_option( self::OPTION_ORDER_STATS_TABLE_HAS_COLUMN_ORDER_FULFILLMENT_STATUS );
+
+		if ( ! empty( $column_status ) ) {
+			return 'yes' === $column_status;
+		}
+
+		global $wpdb;
+
+		$table_name    = self::get_db_table_name();
+		$column_exists = $wpdb->get_var(
+			$wpdb->prepare(
+				// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared -- Table name cannot be prepared.
+				"SHOW COLUMNS FROM `{$table_name}` LIKE %s",
+				'fulfillment_status'
+			)
+		);
+
+		if ( ! empty( $column_exists ) ) {
+			update_option( self::OPTION_ORDER_STATS_TABLE_HAS_COLUMN_ORDER_FULFILLMENT_STATUS, 'yes', false );
+			return true;
+		}
+
+		return false;
+	}
```
Addressed in 20227cb78786c2684fe7a61ee99a30c4e07b06a1

louwie17 commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -1232,6 +1233,45 @@ private static function should_enable_hpos_for_new_shop() {
+	private static function get_order_stats_table_schema( $collate ) {
+		global $wpdb;
+
+		$is_new_install = self::is_new_install();
```
We should also check here if the `fulfillment_status` column already exists or not, given this function would get run on any subsequent DB updates, so a new store may already have the new column, but is not new anymore when a new DB update happens.
This didn't cause any particular warnings for me, but should be addressed.

Nikschavan replied:
I had tested this scenario, dbdelta does not remove anything (indexes, columns) and just adds new changes from schema so this should be safe, that said, I think adding this check and including the column in the schema if site already has that column will make it consistent - I added this in 6fcca417754b260e8cd407d90c66bfab19b6d909

louwie17 commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php:
```diff
@@ -559,6 +539,40 @@ public static function update( $order ) {
+		);
+
+		$order_fulfillment_status = '';
+		if ( true === self::has_fulfillment_status_column() && $order instanceof WC_Order ) {
```
Should we add this check first? `FeaturesUtil::feature_is_enabled( 'fulfillments' )`

Nikschavan replied:
Good call, I added the check in c98efe845dd27ffd528068dda8c0272875be82d5

louwie17 commented on plugins/woocommerce/src/Admin/API/Reports/Orders/Stats/DataStore.php:
```diff
@@ -655,18 +669,51 @@ protected static function get_num_items_sold( $order ) {
+		}
+
+		// Update the option to indicate that the column does not exist.
+		update_option( self::OPTION_ORDER_STATS_TABLE_HAS_COLUMN_ORDER_FULFILLMENT_STATUS, 'no', false );
```
I guess it is ok to leave this permanently set for now until we allow existing stores to migrate?

Nikschavan replied:
Yes, IMO. The debug tool is being added in PR https://github.com/woocommerce/woocommerce/pull/61721. I will rebase and put it for review right after this is merged, so hopefully both are shipped in the same WC release.

In the debug tool, I am also permanently setting this option to 'yes' after the migration is complete - https://github.com/woocommerce/woocommerce/pull/61721/files#diff-7b43719514fca5932b6db23f276afca824fbff5cee5916ab887975deca7ead97R808

louwie17 reviewed (changes requested):
Thanks @Nikschavan, this tested well, and nice work!
I did leave a couple comments and one edge case which didn't break anything in my local environment, but something we did probably want to fix.

Nikschavan committed:
Include the column in schema if it already exists

Nikschavan committed:
Add check if the fulfillments feature is enabled

Nikschavan committed:
Merge branch 'trunk' into new-column-wc-order-stats

Nikschavan committed:
Update test to test with or without column

Nikschavan committed:
use correct option values

louwie17 commented on plugins/woocommerce/includes/class-wc-install.php:
```diff
@@ -1232,6 +1234,48 @@ private static function should_enable_hpos_for_new_shop() {
+	private static function get_order_stats_table_schema( $collate ) {
+		global $wpdb;
+
+		$should_have_fulfillment_column = self::is_new_install();
```
One additional question around this, should we also limit this by the fact the fulfillment feature is enabled? Or is it pretty certain that this will become the default, so having the column there won't hurt to much.
I don't currently have a strong preference btw.

Nikschavan replied:
I can add this in the next PR, I also don't have a strong preferance as well - One negative point is, when a site enables fulfillments after a new install, they wont have this column - but can use the tool

louwie17 reviewed (approved):
This tested well and thanks for the updates!!
I did leave one question, and maybe we should address this as part the follow up debug tool work?
Although assuming fulfillment will be part of core by default, it won't hurt adding the column already for new stores.

Nikschavan commented:
There is no public API change in this PR
